examples/energy-management-app/energy-management-common/include/EnergyEvseTargetsStore.h

```diff
@@ -0,0 +1,135 @@
+#include <lib/core/CHIPError.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/Pool.h>
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <vector>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace EnergyEvse {
+
+class EvseTargetIteratorImpl;
+class EvseTargetsDelegate
+{
+public:
+    using EvseTargetIterator = CommonIterator<EvseTargetEntry>;
+
+    ~EvseTargetsDelegate();
+
+    CHIP_ERROR Init(PersistentStorageDelegate * targetStore);
+    EvseTargetIteratorImpl * GetEvseTargetsIterator();
+    CHIP_ERROR Load(std::vector<EvseTargetEntry> & targetEntryVector, size_t & targetsSize);
+
+    CHIP_ERROR IncreaseEntryCount();
+    CHIP_ERROR DecreaseEntryCount();
+    CHIP_ERROR UpdateEntryCount(bool increase);
+    CHIP_ERROR LoadCounter(size_t & count, size_t & targetsSize);
+    CHIP_ERROR SerializeToTlv(TLV::TLVWriter & writer, const std::vector<EvseTargetEntry> & targetEntryVector);
+
+    /**
+     * @brief   Copies a ChargingTargetSchedule into our store
+     *
+     * @param [in] an entry from the SetTargets list containing:
+     *             dayOfWeekForSequence and chargingTargets (list)
+     *
+     * This routine scans the existing mTargets to see if we have a day of week
+     * set that matches the new target dayOfWeek bits. If there is an existing
+     * matching day then it replaces the days existing targets with the new entry
+     */
+    CHIP_ERROR
+    CopyTarget(const Structs::ChargingTargetScheduleStruct::DecodableType &);
+
+    /**
+     *  @brief   This deletes all targets and resets the list to empty
+     */
+    CHIP_ERROR ClearTargets();
+
+    static constexpr size_t MaxTargetEntryCounterSize()
+    {
+        // All the fields added together
+        return TLV::EstimateStructOverhead(sizeof(size_t), sizeof(size_t));
+    }
+
+    static constexpr size_t MaxTargetEntrySize()
+    {
+        // All the fields added together
+        return TLV::EstimateStructOverhead(sizeof(chip::BitMask<TargetDayOfWeekBitmap>)) +
+            kEvseTargetsMaxNumberOfDays *
+            TLV::EstimateStructOverhead(sizeof(uint16_t), sizeof(Optional<chip::Percent>), sizeof(Optional<int64_t>));
+    }
+
+protected:
+    enum class TargetEntryTag : uint8_t
+    {
+        kTargetEntry           = 1,
+        kDayOfWeek             = 2,
+        kChargingTargetsList   = 3,
+        kChargingTargetsStruct = 4,
+        kTargetTime            = 5,
+        kTargetSoC             = 6,
+        kAddedEnergy           = 7,
+    };
+
+    enum class CounterTag : uint8_t
+    {
+        kCount = 1,
+        kSize  = 2,
+    };
+
+private:
+    // The array itself has a control byte and an end-of-array marker.
+    static constexpr size_t kArrayOverHead = 2;
+
+    EvseTargetIteratorImpl * mEvseTargetsIterator = nullptr;
+    PersistentStorageDelegate * mpTargetStore     = nullptr;
```

Comment:
Suggest looking at the safe persistent storage stuff
 https://github.com/project-chip/connectedhomeip/blob/master/src/app/SafeAttributePersistenceProvider.h

Reply:
I'm already using this for some of the basic attributes, but this data is not stored as an attribute (only accessible as a command Get / Set). Also I don't know if that allows for lists of lists (which is what this needs to have stored).

I wish there was a library that did this, but spoke with others who recommended that it needed to have storage separately implemented so that the storage layer could handle any backwards compatibility changes needed over time. (i.e. independent from the cluster type definitions).

Reply:
Closing this comment out - this code has been refactored since anyway.